staging/src/k8s.io/apiserver/pkg/server/openapi/openapi.go

```diff
@@ -17,35 +17,63 @@ limitations under the License.
 package openapi
 
 import (
+	"bytes"
+	"compress/gzip"
+	"crypto/sha512"
+	"encoding/json"
 	"fmt"
+	"gopkg.in/yaml.v2"
 	"net/http"
 	"reflect"
 	"strings"
 
 	"github.com/emicklei/go-restful"
 	"github.com/go-openapi/spec"
+	"github.com/golang/protobuf/proto"
+	"github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
 
 	"k8s.io/apimachinery/pkg/openapi"
-	"k8s.io/apimachinery/pkg/util/json"
 	genericmux "k8s.io/apiserver/pkg/server/mux"
 	"k8s.io/apiserver/pkg/util/trie"
 )
 
 const (
 	OpenAPIVersion  = "2.0"
 	extensionPrefix = "x-kubernetes-"
+
+	ContentType = "Content-Type"
+	MIME_JSON   = "application/json"
+	MIME_PB     = "application/com.github.googleapis.gnostic.OpenAPIv2+protobuf"
+	MIME_PB_GZ  = "application/x-gzip"
+	MIME_BINARY = "application/octet-stream"
 )
 
 type openAPI struct {
-	config       *openapi.Config
-	swagger      *spec.Swagger
-	protocolList []string
-	servePath    string
-	definitions  map[string]openapi.OpenAPIDefinition
+	config          *openapi.Config
+	swagger         *spec.Swagger
+	swaggerBytes    []byte
+	swaggerPb       []byte
+	swaggerPbGz     []byte
+	swaggerHash     [sha512.Size]byte
+	swaggerPbHash   [sha512.Size]byte
+	swaggerPbGzHash [sha512.Size]byte
+	protocolList    []string
+	servePath       string
+	definitions     map[string]openapi.OpenAPIDefinition
 }
 
 // RegisterOpenAPIService registers a handler to provides standard OpenAPI specification.
+// Note: servePath is the path to the filename with no extension. Supported extensions
+// will be added to the path (e.g. .json, .pb, etc.).
 func RegisterOpenAPIService(servePath string, webServices []*restful.WebService, config *openapi.Config, mux *genericmux.PathRecorderMux) (err error) {
+
+	if !strings.HasSuffix(servePath, ".json") {
+		return fmt.Errorf("Serving path must ends with \".json\".")
```

Comment:
Because OpenAPI spec is in json format. We can relax this limitation if we had a reason for it but right now it helps me remove the .json and add extensions for other formats.